### ISSUE TITLE
Inconsistency regarding generating events

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#524]
**Reported by HansOlsson on 11 Apr 2011 09:56 UTC**
It is not well-defined how non-discrete time non-reals generate events. It is not used much (if at all) but still need to be defined.

Basically the question is how to handle cases like:  
```
model M
  Boolean b=noEvent(time>1) > false;
end M;
```

Background: 3.8.3 state that "a>b" (unless inside noEvent) is a discrete-time expression and will generate events if at least one subexpression is not a discrete-time expression.

The event generation from real elementary relation (see 3.5) is defined in 8.5 and states that at least one operand must be a sub-type of real. There is nothing defined about how other relations generate events.
[Note: Comparison between booleans is defined in 3.5.]

So for
```
  Boolean b=noEvent(time>1) > false;
```
this would imply an event, but unspecified how it is generated.

Possible solutions:
1. Clarify that events are generated also for non-real elementary relations and leave it to tools to do it efficiently (it will be less efficient since no crossing functions can be generated in general). To me it looks odd to first avoid events with noEvent and then automatically get it back later.

2. Change 3.8.3 so that a>b is only discrete if operands are discrete or if it is a real elementary relation generating events as in 3.8.3. This would make 
```
Boolean b=noEvent(time>1) > false;
```
incorrect, since the right-hand side would not be a discrete-time expression.

3. Require that a>b not inside noEvent must be one of: real elementary relation, or comparison between discrete expressions. This would make 
```
 noEvent(time>1) > false
```
incorrect, regardless of where it is used.

When and why was this problem introduced?
In Modelica 1.4, I believe the basic idea was that very early we had
```
  nondiscrete Boolean nb=noEvent(time>1);
  Boolean b=nb > false;
```
in Modelica 1.4 we focused on removing nondiscrete and did not consider the non-discrete boolean expressions in the change. The text in Modelica 1.2/1.3 was close to proposal 2 (but was poorly written in other ways).

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/524

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 23 May 2011 15:11 UTC**
Language group:
Might consider reintroducing nondiscrete in the future.

Poll: Proposal 2:
Favor: 6
Against: 0 
Abstain: 0

Next step is update spec.

**Comment by modelica-trac-importer** [comment]
**Modified by hansolsson on 28 Sep 2011 13:02 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 28 Sep 2011 20:37 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 7 Jul 2015 06:01 UTC**
The nondiscrete keyword is still missing in MLS 3.3. 
Whats about the poll result above? Just to close the ticket is not a solution.

**Comment by modelica-trac-importer** [comment]
**Changelog removed by kurzbach on 7 Jul 2015 06:01 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by kurzbach on 7 Jul 2015 06:01 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 7 Jul 2015 09:20 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 16 Jul 2015 14:13 UTC**
Replying to [comment:4 kurzbach]:
> The nondiscrete keyword is still missing in MLS 3.3. 
> Whats about the poll result above? Just to close the ticket is not a solution.

I understand that this was not clear, and should have clarified it a lot earlier.

The poll above was not about 'nondiscrete' keyword, but about "Proposal 2" from the description of the ticket.

The Proposal 2 in the description states:
"Change 3.8.3 so that a>b is only discrete if operands are discrete or if it is a real elementary relation generating events as in 3.8.3. This would make Boolean b=noEvent(time>1) > false; incorrect, since the right-hand side would not be a discrete-time expression."

This was included in 3.2r1 and 3.3, in the following statement in 3.8.3:
•	Unless inside noEvent: Ordered relations (>,<,>=,<=) if at least one operand is a subtype of Real (i.e. Real elementary relations, see Section ‎3.5) and the functions ceil, floor, div, mod, rem. These will generate events if at least one subexpression is not a discrete-time expression. [In other words, relations inside noEvent(), such as noEvent(x>1), are not discrete-time expressions].


In MLS 3.2 it said:
Unless inside noEvent: Ordered relations (>,<,>=,<=) and the functions ceil, floor, div, mod, rem, abs, sign. These will generate events if at least one subexpression is not a discrete-time expression. [In other words, relations inside noEvent(), such as noEvent(x>1), are not discrete-time expressions].
\--
(Note the subtle difference.)

No decision has been made regarding re-adding a nondiscrete keyword.
